### PR TITLE
Make reverse_indices gpu compatible

### PR DESCRIPTION
**Pull request by yuehhua**
Resolve issue for FluxML/NNlibCUDA.jl#13.

### CODE REVIEW

File: src/utils.jl [hunk 1]
```diff
@@ -15,15 +15,26 @@ maximum_dims(dims::AbstractArray{<:Integer}) = (maximum(dims), )
...
+    # cidx = CartesianIndices(idx)
+    # for i = unique(idx)
+    #     rev[i] = cidx[idx .== i]
+    # end
     rev
 end
 
+function reverse_indices(idx::AbstractArray)
+    rev = Array{Vector{CartesianIndex}}(undef, maximum_dims(idx)...)
+    for i = eachindex(rev)
```
Comment (CarloLucibello) [inline src/utils.jl:32]:
```suggestion
    for i in eachindex(rev)
```

File: src/utils.jl [hunk 2]
```diff
@@ -15,15 +15,26 @@ maximum_dims(dims::AbstractArray{<:Integer}) = (maximum(dims), )
 maximum_dims(dims::AbstractArray{NTuple{N, T}}) where {N,T} = ntuple(i -> maximum(x->x[i], dims), N)
 maximum_dims(dims::AbstractArray{CartesianIndex{N}}) where {N} = ntuple(i -> maximum(x->x[i], dims), N)
 
-function reverse_indices(idx::Array{T}) where T
-    rev = Dict{T,Vector{CartesianIndex}}()
-    for (ind, val) = pairs(idx)
-        rev[val] = get(rev, val, CartesianIndex[])
+function reverse_indices!(rev::AbstractArray, idx::AbstractArray)
+    for (ind, val) = pairs(Array(idx))
```
Comment (CarloLucibello) [inline src/utils.jl:19]:
```suggestion
    for (ind, val) in pairs(Array(idx))
```

File: src/utils.jl [hunk 3]
```diff
@@ -15,15 +15,26 @@ maximum_dims(dims::AbstractArray{<:Integer}) = (maximum(dims), )
...
         push!(rev[val], ind)
     end
+    # if CUDA supports `unique`, a more efficient version:
+    # cidx = CartesianIndices(idx)
+    # for i = unique(idx)
```
Comment (CarloLucibello) [inline src/utils.jl:24]:
```suggestion
    # for i in unique(idx)
```

**Review by CarloLucibello — CHANGES_REQUESTED** [review]

File: src/utils.jl (hunk 3 above)
Comment (yuehhua) [reply src/utils.jl:24]: Done.

File: src/utils.jl (hunk 2 above)
Comment (yuehhua) [reply src/utils.jl]: Done.

File: src/utils.jl (hunk 1 above)
Comment (yuehhua) [reply src/utils.jl]: Done.

**Comment by CarloLucibello** [comment]
It's a pity we have to perform the whole computation on cpu, but I guess we don't have a better solution right now

**Pull request merged**